### PR TITLE
Private endpoint L3 CDK construct

### PR DESCRIPTION
**Pull request by SuperMohit**
## Description

Creates a Private Endpoint in AWS vpcID along with atlas cluster.

Link to any related issue(s): 

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [ ] I have run `make fmt` and formatted my code

### CODE REVIEW

File: cdk/l3-cdk-resources/atlas-private-endpoint/.projenrc.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        'constructs',
+        'cfn-resources',
+        'cloudformation-registry',
+        'l3',
+        'mongodb',
+        'atlas',
+        'cluster',
+        'dbuser',
+        'ip-access-list'],
+    description: 'Creates a project, cluster, dbuser, ipaccess list and private endpoint in MongoDB Atlas',
```
Comment (andreaangiolillo) [inline cdk/l3-cdk-resources/atlas-private-endpoint/.projenrc.js:46]: [small change] you cannot create an ipaccesslist but you add an entry to that list

```suggestion
    description: 'Creates a project, cluster, dbuser, private endpoint and add an IP entry to the access list in MongoDB Atlas',
```

File: cdk/l3-cdk-resources/atlas-private-endpoint/.projenrc.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    // deps: [],                /* Runtime dependencies of this module. */
+    // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
+    // devDeps: [],             /* Build dependencies for this module. */
+    // packageName: undefined,  /* The "name" in package.json. */
```
Comment (andreaangiolillo) [inline cdk/l3-cdk-resources/atlas-private-endpoint/.projenrc.js:51]:
```suggestion
```

File: cdk/l3-cdk-resources/atlas-private-endpoint/.projenrc.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+new JsonFile(project, 'cdk.json', {
+    obj: {
+        app: 'npx ts-node --prefer-ts-exts src/integ.default.ts',
+    },
+});
+
+const common_exclude = ['cdk.out', 'cdk.context.json', 'yarn-error.log'];
+project.npmignore.exclude(...common_exclude);
+project.gitignore.exclude(...common_exclude);
```
Comment (andreaangiolillo) [inline cdk/l3-cdk-resources/atlas-private-endpoint/.projenrc.js:62]: Why is this needed?

File: cdk/l3-cdk-resources/atlas-private-endpoint/README.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+The official [MongoDB Atlas](https://www.mongodb.com/) AWS CDK resource for Node.js.
+
+> This construct uses MongoDB [L1 construct](https://constructs.dev/search?q=&offset=0&tags=mongodb-published) and data structures for the [AWS CloudFormation Registry] L3 type.
+
+[L1 construct]: https://docs.aws.amazon.com/cdk/latest/guide/constructs.html
+[AWS CloudFormation Registry]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/registry.html
+
+## Description
+
+This construct lets you create Project, Cluster, DB User and configures IpAccesslist and Private Endpoint.
```
Comment (andreaangiolillo) [inline cdk/l3-cdk-resources/atlas-private-endpoint/README.md:12]: same comment for the ip access list

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+ * @export
+ * @interface AtlasPrivateEndpointProps
+ */
+export interface AtlasPrivateEndpointProps {
+    /**
+     * @description
+     * @type {string}
+     * @memberof AtlasPrivateEndpointProps
+     */
+    readonly groupId?: string;
```
Comment (andreaangiolillo) [inline cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts:77]: Let's use `projectId`. we should stay consistent here, in terraform and cli we use `projectId` . 
```suggestion
    readonly projectId?: string;
```

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+     * @type {string}
+     * @memberof PrivateEndpointProps
+     */
+    readonly status?: string;
+    /**
+     * @description
+     * @type {string}
+     * @memberof PrivateEndpointProps
+     */
+    readonly groupId?: string;
```
Comment (andreaangiolillo) [inline cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts:134]:
```suggestion
    readonly projectId?: string;
```

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+     * @type {ApiKey}
+     * @memberof PrivateEndpointProps
+     */
+    readonly apiKeys?: ApiKey;
+    /**
+     * @description
+     * @type {string}
+     * @memberof PrivateEndpointProps
+     */
+    readonly region?: string;
```
Comment (andreaangiolillo) [inline cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts:146]: can we set a defaul value for the region if not provided? `us-east-1` is a good candidate

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    readonly private_endpoint : CfnPrivateEndpoint;
+
+    /**
+     * Creates an instance of AtlasBasicPrivateEndpoint.
+     * @param {Construct} scope
+     * @param {string} id
+     * @param {AtlasPrivateEndpointProps} props
+     * @memberof AtlasBasicPrivateEndpoint
+     */
+    constructor(scope: Construct, id: string, props: AtlasPrivateEndpointProps) {
```
Comment (andreaangiolillo) [inline cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts:47]: can we add validation of the input so that the constructor will throw an exception if a required field is undefined or empty string? we want to fail fast so that the user does not have to wait for the stack to run to get an error

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+     * @type {string}
+     * @memberof AtlasPrivateEndpointProps
+     */
+    readonly groupId?: string;
+    /**
+     * @description
+     * @type {ApiKeyDefinition}
+     * @memberof AtlasPrivateEndpointProps
+     */
+    readonly apiKeys: ApiKeyDefinition;
```
Comment (andreaangiolillo) [inline cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts:83]: Can we have the public and private keys as fields here instead of using `ApiKeyDefinition`?

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+     * @memberof AtlasBasicPrivateEndpoint
+     */
+    constructor(scope: Construct, id: string, props: AtlasPrivateEndpointProps) {
+        super(scope, id);
+        
+        this.atlas = new AtlasBasic(this, 'atlas-basic-'.concat(id),
+            {
+                apiKeys: props.apiKeys,
+                ...props.atlasBasicProps,
+            });
```
Comment (andreaangiolillo) [inline cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts:54]: what atlas stands for and why `private_endpoint` is not part of `atlas`? `atlas` is too generic

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+     */
+    readonly privateEndpointProps: PrivateEndpointProps;
+}
+
+/**
+ * @description
+ * @export
+ * @interface PrivateEndpointProps
+ */
+export interface PrivateEndpointProps {
```
Comment (andreaangiolillo) [inline cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts:110]: I see another `PrivateEndpointProps` at line 71. Is this correct? what do we have two `PrivateEndpointProps`?

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/integ.default.ts [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        subnetIds: [subnetId]
+    }],
+}
+
+const props   = {
+    apiKeys: apiKeys,
+    atlasBasicProps: atlasBasicProps,
+    privateEndpointProps: privateEndpointProps,
+    region: awsRegion
+}
```
Comment (andreaangiolillo) [inline cdk/l3-cdk-resources/atlas-private-endpoint/src/integ.default.ts:80]: [q] Can we just define a single interface with all the properties instead of asking the user to create 3 different interfaces?

Comment (SuperMohit) [reply cdk/l3-cdk-resources/atlas-private-endpoint/src/integ.default.ts:80]: [a] props is a single interface. The user is free to populate it as he likes. It is an app code.

File: cdk/l3-cdk-resources/atlas-private-endpoint/.projenrc.js (hunk 3 above)
Comment (SuperMohit) [reply cdk/l3-cdk-resources/atlas-private-endpoint/.projenrc.js:62]: excludes this file from packaging.

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts (hunk 7 above)
Comment (SuperMohit) [reply cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts:146]: Yes, It's already set.

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts (hunk 11 above)
Comment (SuperMohit) [reply cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts:110]: No there is only 1.

File: cdk/l3-cdk-resources/atlas-private-endpoint/.projenrc.js (hunk 3 above)
Comment (andreaangiolillo) [reply cdk/l3-cdk-resources/atlas-private-endpoint/.projenrc.js:62]: ah, thanks!
what about this?
```
new JsonFile(project, 'cdk.json', {
    obj: {
        app: 'npx ts-node --prefer-ts-exts src/integ.default.ts',
    },
});
```

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts (hunk 9 above)
Comment (SuperMohit) [reply cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts]: updated.

**Comment by SuperMohit** [comment]
Tested Private Endpoint :
Screenshot 2023-02-10 at 11 04 19 PM
Screenshot 2023-02-10 at 11 04 34 PM

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts (hunk 8 above)
Comment (SuperMohit) [reply cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts:45]: added the validation. for AtlasBasic it would be updated in the next release.

File: cdk/l3-cdk-resources/atlas-private-endpoint/.projenrc.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    authorName: 'MongoDBAtlas',
+    cdkVersion: '2.1.0',
+    defaultReleaseBranch: 'main',
+    name: '@mongodbatlas-awscdk/atlas-basic-private-endpoint',
+    repositoryUrl: 'https://github.com/mongodb/mongodbatlas-cloudformation-resources.git',
+    stability: 'experimental',
+    peerDeps: [
+        '@mongodbatlas-awscdk/atlas-basic',
+        '@mongodbatlas-awscdk/private-endpoint'
+    ],
```
[the remark this answers is not in the document]
Comment (SuperMohit) [reply cdk/l3-cdk-resources/atlas-private-endpoint/.projenrc.js]: https://docs.aws.amazon.com/cdk/v2/guide/manage-dependencies.html

File: cdk/l3-cdk-resources/atlas-private-endpoint/README.md [hunk 14]
```diff
@@ -0,0 +1,166 @@
+# @mongodbatlas-awscdk/atlas-basic-private-endpoint
+
+The official [MongoDB Atlas](https://www.mongodb.com/) AWS CDK resource for Node.js.
+
+> This construct uses MongoDB [L1 construct](https://constructs.dev/search?q=&offset=0&tags=mongodb-published) and data structures for the [AWS CloudFormation Registry] L3 type.
```
Comment (Zuhairahmed) [inline cdk/l3-cdk-resources/atlas-private-endpoint/README.md:5]:
```suggestion
> This construct uses MongoDB [Level 1 construct](https://constructs.dev/search?q=&offset=0&tags=mongodb-published) and data structures for the [AWS CloudFormation Registry] Level 3 type.
```

File: cdk/l3-cdk-resources/atlas-private-endpoint/README.md [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+
+[L1 construct]: https://docs.aws.amazon.com/cdk/latest/guide/constructs.html
+[AWS CloudFormation Registry]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/registry.html
+
+## Description
+
+Creates a project, cluster, dbuser, private endpoint and add an IP entry to the access list in MongoDB Atlas.
```
Comment (Zuhairahmed) [inline cdk/l3-cdk-resources/atlas-private-endpoint/README.md:12]:
```suggestion
Creates a MongoDB Atlas Project, Cluster, DBuser, Private Endpoint and adds an IP entry to the IP Access List.
```

File: cdk/l3-cdk-resources/atlas-private-endpoint/README.md [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+
+
+## MongoDB Atlas API Docs
+
+For more information about the API refer to: [API Endpoints](https://www.mongodb.com/docs/atlas/reference/api-resources-spec)
```
Comment (Zuhairahmed) [inline cdk/l3-cdk-resources/atlas-private-endpoint/README.md:17]:
```suggestion
For more information for each of the specific APIs used in this Level 3 AWS CDK resource please refer to: [Official Documentation](https://www.mongodb.com/docs/atlas/reference/api-resources-spec)
```

File: cdk/l3-cdk-resources/atlas-private-endpoint/README.md [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+    apiKeys: apiKeys,
+    atlasBasicProps: atlasBasicProps,
+    privateEndpointProps: privateEndpointProps,
+    region: awsRegion
+}
+
+new AtlasBasicPrivateEndpoint(stack,'private-endpoint', props)
+```
+
+Default Region is set to us-east-1
```
Comment (Zuhairahmed) [inline cdk/l3-cdk-resources/atlas-private-endpoint/README.md:152]:
```suggestion
Default Region is set to us-east-1 region (AWS US East N. Virginia)
```

File: cdk/l3-cdk-resources/atlas-private-endpoint/README.md [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+
+* Issues related to this generated library should be [reported here](https://github.com/cdklabs/cdk-cloudformation/issues/new?title=Issue+with+%40cdk-cloudformation%2Fmongodb-atlas-cluster+v1.0.0).
+* Issues related to this library should be reported to the [publisher](https://github.com/mongodb/mongodbatlas-cloudformation-resources/issues).
+* Feature requests should be [reported here](https://feedback.mongodb.com/forums/924145-atlas?category_id=392596)
+
+[cdklabs/cdk-cloudformation]: https://github.com/cdklabs/cdk-cloudformation
+
+## License
+
+Distributed under the Apache-2.0 License.# replace this
```
Comment (Zuhairahmed) [inline cdk/l3-cdk-resources/atlas-private-endpoint/README.md:166]: is "replace this" a typo or is this a note to add something else here?

**Review by Zuhairahmed — APPROVED** [review]
few minor comments, non-blocking otherwise LGTM

**Comment by Zuhairahmed** [comment]
@SuperMohit can you move this to master branch? just FYI @andreaangiolillo we can move to dedicated CDK repo soon, but for time being its hard for customers to find+review source code if we are in production in a branch other than master

File: cdk/l3-cdk-resources/atlas-private-endpoint/README.md [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+    privateEndpointProps: privateEndpointProps,
+    region: awsRegion
+}
+
+new AtlasBasicPrivateEndpoint(stack,'private-endpoint', props)
+```
+
+Default Region is set to us-east-1
+
+You can find more information about activating this type in the [AWS CloudFormation documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/registry-public.html).
```
Comment (Zuhairahmed) [inline cdk/l3-cdk-resources/atlas-private-endpoint/README.md:154]: @SuperMohit can you add in additional section for L1 defaults in library assumed like you did here: https://github.com/mongodb/mongodbatlas-cloudformation-resources/blob/INTMDB-548/cdk/l3-cdk-resources/atlas-basic/README.md

Comment (Zuhairahmed) [reply cdk/l3-cdk-resources/atlas-private-endpoint/README.md:154]: Also to make IP Access List open to who world we to both CIDRs: ipAddress: '0.0.0.0/1' and '128.0.0.0/1'

Comment (SuperMohit) [reply cdk/l3-cdk-resources/atlas-private-endpoint/README.md:154]: I don't even want it to open for the first half. Don't think it is the right approach for the default. Default should be `deny all`

**Comment by andreaangiolillo** [comment]
@SuperMohit Could you fix shellcheck?

**Comment by SuperMohit** [comment]
> @SuperMohit can you move this to master branch? just FYI @andreaangiolillo we can move to dedicated aws-cdk repo soon, but for time being its hard for customers to find+review source code if we are in production in a branch other than master

Yes, I have raised the PR to change the base branch. Once that is merged I can move this PR as well.

**Review by andreaangiolillo — APPROVED** [review]
LGTM

File: cdk/l3-cdk-resources/atlas-private-endpoint/README.md (hunk 19 above)
Comment (Zuhairahmed) [reply cdk/l3-cdk-resources/atlas-private-endpoint/README.md:185]: makes sense, if we go with `deny all` then make sure this is clearly called out in your README file

**Comment by colm-quinn** [comment]
Bump to see can we merge these changes.

File: cdk/l3-cdk-resources/atlas-private-endpoint/README.md [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+  --type RESOURCE \
+  --execution-role-arn ROLE-ARN  
+  
+aws cloudformation activate-type \
+  --type-name MongoDB::Atlas::PrivateEndpoint \
+  --publisher-id bb989456c78c398a858fef18f2ca1bfc1fbba082 \
+  --type RESOURCE \
+  --execution-role-arn ROLE-ARN
+```
+Alternatively:
```
Comment (colm-quinn) [inline cdk/l3-cdk-resources/atlas-private-endpoint/README.md:54]: [nit] Could you add context on why the alternative command is required here? It's a little confusing

File: cdk/l3-cdk-resources/atlas-private-endpoint/README.md [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+const app = new cdk.App();
+
+const stack = new cdk.Stack(app, 'atlas-basic-default', {
+    env: { region: process.env.CDK_DEFAULT_REGION, account: process.env.CDK_DEFAULT_ACCOUNT },
+});
+
+const apiKeys = {
+    privateKey: stack.node.tryGetContext('MONGODB_ATLAS_PRIVATE_KEY') || process.env.MONGODB_ATLAS_PRIVATE_KEY,
+    publicKey: stack.node.tryGetContext('MONGODB_ATLAS_PUBLIC_KEY') || process.env.MONGODB_ATLAS_PUBLIC_KEY,
+};
```
Comment (colm-quinn) [inline cdk/l3-cdk-resources/atlas-private-endpoint/README.md:96]: Out of date - we now use the profile correct?

File: cdk/l3-cdk-resources/atlas-private-endpoint/README.md (hunk 20 above)
Comment (SuperMohit) [reply cdk/l3-cdk-resources/atlas-private-endpoint/README.md]: I have removed it.

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+
+
+/**
+ * @description
+ * @export
+ * @interface AtlasPrivateEndpointProps
+ */
+export interface AtlasPrivateEndpointProps {
+
+    readonly profile?: string;
```
Comment (maastha) [inline cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts:76]: This shouldn't be optional right? Or am I missing something?

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/integ.default.ts [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+
+const privateEndpointProps : PrivateEndpointProps = {
+    privateEndpoints: [{
+        vpcId: vpcId,
+        subnetIds: [subnetId]
+    }],
+}
+
+const props   = {
```
Comment (maastha) [inline cdk/l3-cdk-resources/atlas-private-endpoint/src/integ.default.ts:77]: profile?

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+    readonly privateEndpoint : CfnPrivateEndpoint;
+
+    /**
+     * Creates an instance of AtlasBasicPrivateEndpoint.
+     * @param {Construct} scope
+     * @param {string} id
+     * @param {AtlasPrivateEndpointProps} props
+     * @memberof AtlasBasicPrivateEndpoint
+     */
+    constructor(scope: Construct, id: string, props: AtlasPrivateEndpointProps) {
```
Comment (maastha) [inline cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts:45]: [nit] This can be called AtlasBasicPrivateEndpointProps for consistency.

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts (hunk 22 above)
Comment (SuperMohit) [reply cdk/l3-cdk-resources/atlas-private-endpoint/src/index.ts:76]: profile is optional everywhere. Any resource we have not set the default?

File: cdk/l3-cdk-resources/atlas-private-endpoint/src/integ.default.ts (hunk 23 above)
Comment (SuperMohit) [reply cdk/l3-cdk-resources/atlas-private-endpoint/src/integ.default.ts:77]: profile is optional.

**Review by maastha — APPROVED** [review]
LGTM
